enhancements/cluster-logging/loki-storage.md

```diff
@@ -0,0 +1,623 @@
+
+**Mitigation:** Limit by maintaining a strong opinionated API for Loki deployments that encompasses only features around dependencies (e.g. object storage), scheduling (e.g. replicas, tolerations/node-selectors) and no more.
+
+**Risk:** Providing poor user experience for day 1 and day 2, e.g. user needs to figure out requests/limits for each component.
+
+**Mitigation:** Provide opinionated and battle-tested scale out formats in form of a T-Shirt size (e.g. `1x.medium`). Leaving room for adding more formats later (e.g. `2x.medium` for hard multi-tenancy) and knobs (e.g. replicas) to adapt the format to live environments.
+
+## Design Details
+
+The discussion about the proposed API is split in two sections. The first section `lokistack_types.go` describes the custom resource definition for the [loki-operator](https://github.com/viaq/loki-operator) to reconcile a Loki deployment.
+
+The API in `lokistack_types.go` includes all required and optional configuration options to support the [proposal](#proposal) above. Furthermore the status struct in `lokistack_types.go` declares conditions for the Loki deployment. The Loki deployment can be under various situations in a degraded state resulting in no deployment.
+
+The second section `clusterlogging_types.go` describes the minimal stanza for translating the `logStore` spec into a `LokiStackSpec`.
+
+### lokistack_types.go
+
+#### Types and constants
+
+A `LokiStack` custom resource can declare one of the following three sizes as values of the `LokiStackSpec`'s field `Size`. Each value has impact to all of the following per Loki component:
+
+1. `Replicas`: The number of replica pods per Loki component. Besides two exceptions all components are deployed with at minimum two replicas. First exception is the `1x.extra-small` resulting in a single replica per component. Second exception is the Loki compactor that is always deployed with a single replica, because multiple replicas are not supported.
+2. `Requests/Limits`: The number of requested CPU (in millicores) and Memory (in Bytes) and their limits per component.
+3. `PodDisruptionBudgets`: The number of minimum available pods per component to ensure normal operations under selected settings, e.g. `ReplicationFactor = 1`.
+
+```go
+// ManagementStateType defines the type for CR management states.
+//
+// +kubebuilder:validation:Enum=Managed;Unmanaged
+type ManagementStateType string
+
+const (
+    // ManagementStateManaged when the LokiStack custom resource should be
+    // reconciled by the operator.
+    ManagementStateManaged ManagementStateType = "Managed"
+
+    // ManagementStateUnmanaged when the LokiStack custom resource should not be
+    // reconciled by the operator.
+    ManagementStateUnmanaged ManagementStateType = "Unmanaged"
+)
+
+// LokiStackSizeType declares the type for Loki deployment scale outs.
+//
+// +kubebuilder:validation:Enum=OneXExtraSmallSize;OneXSmall;OneXMedium
```

Comment:
I wonder If it's intentional that `OneXExtraSmallSize` has **size** but not others `OneXSmall;OneXMedium` fields.

Reply:
@vishnoisuresh  This is a typo! Thanks for watching out! As per source code in enhancement proposal being more for demonstration, what matters is on the operator source code. AFAICS, we have this fixed there:
```
// +kubebuilder:validation:Enum=SizeOneXExtraSmall;SizeOneXSmall;SizeOneXMedium
```